### PR TITLE
docs: tutorial

### LINKED ISSUE #122: Revamp get started tutorial
A consistent feedback we have received from multiple places is that developers have a hard time finding information on our docs. Sometimes it is becacuse that documentation does not exist, others because it is not easily discoverable (how to search for what you don't know?).

The following is a WIP outline of the Quick start guide, so feedback is welcome:

- Introduction
  - Web concepts
  - nodejs installed
  - [Process model](https://www.electronjs.org/docs/latest/tutorial/process-model)
- Scaffolding
  - [Debugging with VS Code](https://www.electronjs.org/docs/latest/tutorial/debugging-vscode) --> Maybe add the configuration to the quick-start?
- Accessing Node.js APIs--> Talk about [Context isolation](https://www.electronjs.org/docs/latest/tutorial/context-isolation) and prerenderer and how to implement things.
  - Link again to [process model](https://www.electronjs.org/docs/latest/tutorial/process-model) or mention things?
  - Use `process.versions` as an example on how to implement, mention that node integration is to be avoided
  - Link to the security docs ([Context isolation](https://www.electronjs.org/docs/latest/tutorial/context-isolation), [Process sandboxing](https://www.electronjs.org/docs/latest/tutorial/sandbox), [Checklist](https://www.electronjs.org/docs/latest/tutorial/security))
- Deeper integration with the OS
  - Have a list of all the other stuff --> Manually for now, but will be good to get something populated automatically that uses the description and other stuff
- [Application distribution](https://www.electronjs.org/docs/latest/tutorial/application-distribution)
  - Link to all the other links under distribution ([Code signing](https://www.electronjs.org/docs/latest/tutorial/code-signing), [Mac App Store Submission Guide](https://www.electronjs.org/docs/latest/tutorial/mac-app-store-submission-guide), [Windows Store Guide](https://www.electronjs.org/docs/latest/tutorial/windows-store-guide), [Snapcraft Guide (Linux)](https://www.electronjs.org/docs/latest/tutorial/snapcraft)
- [Updating Applications](https://www.electronjs.org/docs/latest/tutorial/updates)
- Further reading: Go to Examples and look at the API, [performance](https://www.electronjs.org/docs/latest/tutorial/performance), [fuses](https://www.electronjs.org/docs/latest/tutorial/fuses), etc.?

### PR DESCRIPTION
**Pull request by molant**
This is a contentent reorganization + expand and split the tutorial to make it easier for users to find the right information.

I'm opening as a draft to avoid merging it as all these changes will have to happen upstream so they are not reverted in the next docs automatic update.

There might be typos and such, but this should be ready to start discussing.

Fix #122

### CODE REVIEW

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    }
+  ]
+}
+```
+
+3. The option `Debug Electron's Main Process` will appear when you select `Run and Debug`
+   from the sidebar, allowing you to set breakpoints and inspect all the variables among
+   other things.
+
+![Screenshot of VS Code debugging the code so far](/assets/img/vscode-debugger.png)
```
[the remark this answers is not in the document]
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-scaffolding.md]: I mean the entire VSCode debugger part

File: docs/latest/tutorial/tutorial-node-access.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+At this moment you have code being executed in the main and renderer process.
+The easiest way to debug the code in your renderer process is by using the built-in developer
+tools. You can invoke them with a shortcut (`Ctrl/Command+Shift+I`) or programmatically
+via `mainWindow.webContents.openDevTools()` in the main process.
+
+If you want to use VS Code to debug the main and renderer process simulatenously from the
+same tool, you have to use a [compound task].
+To do so, open the file `.vscode/launch.json` that you created in the previous step and update
+it to look as follows (changes are highlighted):
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-node-access.md:167]: Don't know if this fits into the `Accessing Node.js APIs` section of the tutorial

File: docs/latest/tutorial/tutorial-os-integration.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+:::
+
+There are many ways to integrate with the Operating System. In some cases it could be platform specific
+(like Apple's TouchBar), in others it could be through APIs to have a more integrated look and feel
+(like modifying the Title Bar), or using APIs that give you access to USB and Bluetooth devices.
+
+The following is a list of all the guides currently available. If there is still something that you do
+not know how to do, please take a look at the [API][app] as there is a chance it might be documented
+just there (and also open an issue requesting the guide!).
```
[the remark this answers is not in the document]
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-os-integration.md:27]: > I think that repeating content is fine, isn't there a say such as "when you are sick of repeating yourself, that's when people start listening to you"? or something around those lines? We have had this content for a longtime yet people are having trouble finding it.

My only issue with repetition in writing is that it makes it easier for docs to go stale. I think making it easier for people to find the single place where the information is stored would be the way to go.

IMO the best way would be to go about doing this would be a few different steps:

1) Split off the "How-To" category in the sidebar into a "Desktop Environment Integration" category and... one or more other ones (TBD).

1) The "Desktop Environment Integration" section should have an "Overview" doc that contains this content.

1) Rename the `Managing your window's lifecycle` to `Desktop environment integration` or `Basic OS integration` because making the window lifecycle match OS-specific behaviour is part of that broader category, and add a note redirecting to the Overview doc from the last point.

I think a re-org like this would make more sense because you can find info in the same spots, but we're repeating a bit less.

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md]: I think the whole section would look weird as an admonition unless we split it into another file. The reality is that the vast majority of Electron developers are using VS Code so I'd just make this information as visible as possible (and if we ever create custom boilerplate I'd add those files as well).

File: docs/latest/tutorial/tutorial-node-access.md (hunk 2 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-node-access.md:167]: What do you think of adding some JS in [Adding some UI](https://electronjsorg-new-pr-124.herokuapp.com/docs/latest/tutorial/scaffolding#adding-some-ui) and add this debugging part there as well? That way all the VS Code debugging is in the same section.

File: docs/latest/tutorial/tutorial-os-integration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-os-integration.md:27]: > My only issue with repetition in writing is that it makes it easier for docs to go stale. 

Yes, we need to do something about all the stale ones we have already 😔

I like your proposal!

> Split off the "How-To" category in the sidebar into a "Desktop Environment Integration" category and... one or more other ones (TBD).

What about:

- OS integration (thinking about the space we have in the sidebar)
- How to (just keep the rest there)

[the remark this answers is not in the document]
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-os-integration.md:27]: > What about:

> OS integration (thinking about the space we have in the sidebar)
> How to (just keep the rest there)

I like that!

File: docs/latest/tutorial/tutorial-os-integration.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+| [File drag & drop]     | Use the OS' drag & drop feature to content into/from your application                            |
+| [Notifications]        | Send notifications to your users from the main and renderer processes                            |
+| [Online/Offline]       | Detect if your application has Internet access and behave accordingly                            |
+| [Taskbar progress bar] | Show progress information to your users directly from the taskbar of the OS                      |
+| [Recent documents]     | Provide access to a list of recent documents opened by the application via JumpList or dock menu |
+| [Represented file]     | On macOS, you can set a represented file for any window in your application                      |
+| [Spellchecker]         | Learn how to use the built-in spellchecker, set languages, etc.                                  |
+| [Tray]                 | Add a try icon to your application with its own context menu                                     |
+| [Windows taskbar]      | Customize your app's Windows taskbar with Jumplist, thumbnails and toolbars, icon overlays, etc. |
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-os-integration.md]: I've created a new [prebuild step](https://github.com/electron/electronjs.org-new/pull/124/files#diff-57a3a0322be9a0ad77ec67fb0eba5f6eb289b2034e72b07546d72324158c3e93) that takes all the items in that category from sidebars.js and updates the table automatically.

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md]: I've grouped all the VS Code debugging and put it into the scaffolding part

File: docs/latest/tutorial/tutorial-os-integration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-os-integration.md]: @erickzhao do we want to link to OS Integration or Examples from the sidebar?

File: docs/latest/tutorial/tutorial-node-access.md (hunk 2 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-node-access.md]: Moved all the debugging and VS Code stuff to the scaffolding step

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+```js title='main.js'
+const { app, BrowserWindow } = require('electron')
+
+const createWindow = () => {
+  const win = new BrowserWindow({
+    width: 800,
+    height: 600,
+    webPreferences: {
+      sandbox: true,
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:167]: I think we should align with `electron-quick-start` for props here. Not sure if we should be including `sandbox` without greater alignment around `sandbox: true` across the docs, or at least more explanation on what the prop does here and linking to the Sandboxing doc.

Also unrelated but there's a trailing comma here but none in the next instance of this snippet.

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+- The [`app`][app] module, which controls your application's event lifecycle.
+- The [`BrowserWindow`][browser-window] module, which creates and manages application
+  windows.
+
+```js
+const { app, BrowserWindow } = require('electron')
+```
+
+:::warning
+ESM (i.e.: use `import` to load a module) is currently not supported in Electron directly.
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:190]:
```suggestion
[ECMAScript modules](https://nodejs.org/api/esm.html) (i.e. using `import` to load a module) is currently not supported in Electron directly.
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+      sandbox: true
+    }
+  })
+
+  win.loadFile('index.html')
+}
+```
+
+:::tip
+You can find all the options available in the [`BrowserWindow` page][browser-window].
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:213]:
```suggestion
You can find all the constructor options available in the [`BrowserWindow` API docs][browser-window].
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+In your console you should read `Hello from Electron 👋`.
+
+Congratulations, you have executed your first Electron application!
+
+:::tip
+If you haven't yet, please read [Electron's process model][process-model] to better understand
+how all the pieces work together.
+:::
+
+## Adding some UI
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:124]:
```suggestion
## Creating a window
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+resolves its Promise. This is done in the last 3 lines of the code:
+
+```js
+app.whenReady().then(() => {
+  createWindow()
+})
+```
+
+At this point, your Electron application should successfully open a window that displays your
+web page!
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:228]: I realize now that we don't really introduce the concept of the renderer process at this point:

```suggestion
At this point, your Electron application should successfully open a window that displays your
web page! Each window created by your app will run in a separate process called a
**renderer**.
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+
+You have now code being executed in the main and renderer process.
+The easiest way to debug the code in your renderer process is by using the built-in developer
+tools. You can invoke them with a shortcut (`Ctrl/Command+Shift+I`) or programmatically
+via `mainWindow.webContents.openDevTools()` in the main process. If you want to use VS Code,
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:233]:
```suggestion
via `win.webContents.openDevTools()` in the main process. If you want to use VS Code,
```
Aligning the window reference variable name here

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+
+The renderer has access to the web APIs. That means you can use the same JavaScript APIs and
+tooling you use for typical front-end development, such as using [`webpack`][webpack] to bundle
+and minify your code or [React][react] to manage your user interfaces.
+
+:::tip
+You can find more information about this subject in the [Boilerplates and CLIs].
+:::
+
+### Debugging from VS Code
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:244]: Is this meant to be an H3? I feel it'd fit better as an H2.

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+      "outputCapture": "std",
+      "console": "integratedTerminal"
+    }
+  ]
+}
+```
+
+<!-- markdownlint-disable MD029 -->
+
+3. The option `Main + renderer` will appear when you select `Run and Debug`
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:291]:
```suggestion
1. The option `Main + renderer` will appear when you select `Run and Debug`
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+from the sidebar, allowing you to set breakpoints and inspect all the variables among
+other things in both the main and renderer processes.
+<!-- markdownlint-enable MD029 -->
+
+What we have done in the `lauch.json` file is to create 3 configurations:
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:296]:
```suggestion
What we have done in the `launch.json` file is to create 3 configurations:
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+
+- `Main` is use to start the main process and also expose the port 9222 for remote debugging
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:298]:
```suggestion
- `Main` is used to start the main process and also expose port 9222 for remote debugging
```

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 11 above)
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-scaffolding.md:244]:
```suggestion
### Optional: Debugging from VS Code
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 15]
```diff
@@ -0,0 +1,354 @@
...
+:::caution
+Because we are attaching to a process that exists already in the `Renderer` configuration
+(`"request": "attach"`) there is a high chance that you will not be able to set a breakpoint
+in the first lines of your code as the debugger will not have had enough time to connect before
+they are being executed.
+You can workaround this by refreshing the page or setting a timeout before executing the code
+in development mode.
+:::
+
+:::tip Further reading
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:315]: Should we add titles to all the other admonitions as well? :)

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 16]
```diff
@@ -0,0 +1,354 @@
...
+  (`--remote-debugging-port=9222`). This is the port that we will use to attach the debugger
+  for the `Renderer`. Because the main process is a Node.js process, the type is set to
+  `pwa-node` (`pwa-` is the prefix that tells VS Code to use the latest JavaScript debugger).
+- `Renderer` is used to debug the renderer process. It tries to connect to the port 9222.
+  Because the renderer is a web process, the debugger that we use is `pwa-chrome`.
+- `Main + renderer` is a [compound task] that executes the previous ones simulatenously.
+
+:::caution
+Because we are attaching to a process that exists already in the `Renderer` configuration
+(`"request": "attach"`) there is a high chance that you will not be able to set a breakpoint
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:308]: I think we should explain the `request: attach` thing above. I'm not too familiar with it myself. Does it mean that we're attaching to the existing remote debugging port at 9222 created when running the "Main" configuration?

Does this also work with multiple renderer processes?

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+in the first lines of your code as the debugger will not have had enough time to connect before
+they are being executed.
+You can workaround this by refreshing the page or setting a timeout before executing the code
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-scaffolding.md:311]:
```suggestion
You can work around this by refreshing the page or setting a timeout before executing the code
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+}
+```
+
+This `start` command will let you open your app in development mode.
+
+```sh npm2yarn
+npm start
+```
+
+:::tip
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:88]:
```suggestion
:::info
```

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 15 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:315]: There are some that make sense, others with the default "Note" or "Tip" I think are good. I'll review them just in case.

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 19]
```diff
@@ -0,0 +1,189 @@
+---
+title: 'Main and Renderer process communication'
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]:
```suggestion
title: 'main and renderer process communication'
```

File: docs/latest/tutorial/tutorial-scaffolding.md [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+To initialize the `main` script, create a file named `main.js` in the root folder
+of your project with the following contents and run `npm start` again:
+
+```js title='main.js'
+console.log(`Hello from Electron 👋`);
+```
+
+In your console you should read `Hello from Electron 👋`.
+
+Congratulations, you have executed your first Electron application!
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:117]: But it's true! Need to keep them motivated :D

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 5 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:167]: weird that the comma isn't failing in the pr to e/e.

I'd rather use `sandbox` in the tutorial. That's what we want people to use and the default in this cases should be `true` soon-ish when not specified.

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 16 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:308]: > Does it mean that we're attaching to the existing remote debugging port at 9222 created when running the "Main" configuration?

Yes, haven't tried the multiple renderer process.
I'll make it clearer.

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 5 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:167]:
```suggestion
      sandbox: true
```

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 11 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:244]:
```suggestion
## Optional: Debugging from VS Code
```

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 12 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md:291]: It has to be a 3, otherwise the numbering starts again at 1. That's why I have disabled the markdownlint rule

File: docs/latest/tutorial/tutorial-prerequisites.md [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+:::info Tutorial parts
+This is part 1 of the Electron tutorial. The other parts are:
+
+- [Prerequisites]
+- [Scaffolding][scaffolding]
+- [Main and Renderer process communication][main-renderer]
+- [Adding features][features]
+- [Application distribution]
+- [Code signing]
+- [Updating your application][updates]
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/tutorial-prerequisites.md]: Updated it

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 12 above)
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-scaffolding.md:291]: Oh I see! I missed that.

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+1. [Main and Renderer process communication][main-renderer]
+1. [Adding Features][features]
+1. [Application Distribution]
+1. [Code Signing]
+1. [Updating Your Application][updates]
+
+:::
+
+At the end of this part you will have an Electron application that uses
+a `preload` script to safely expose features from the main process
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:22]: Open to feedback here, but it might make sense to not code fence `preload` during normal usage because we use the term a lot.

```suggestion
a preload script to safely expose features from the main process
```

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+into the renderer one.
+
+## Main and Renderer process communication
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:25]: Don't really like this heading because it repeats the title

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 24]
```diff
@@ -0,0 +1,189 @@
+---
+title: 'Main and Renderer Process Communication'
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:2]: I don't think this title is super accurate because the preload script doesn't actually communicate with the main process. There's no IPC involved here AFAIK, the preload script is just privileged and has access to some Node.js APIs that the renderer process does.

I'd maybe rename this to `Using a preload script` or something similar.

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+This is your last reminder to look at [Electron's Process Model][process-model] to
+better understand all of this!
+:::
+
+This is where attaching a [`preload` script][preload-script] to your
+renderer comes in handy. A `preload` script runs before the renderer process is loaded,
+and has access to both renderer globals (e.g. `window` and `document`) and a Node.js
+environment.
+
+We are going to creae a `preload` script that exposes the versions of Chrome, Node, and
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:51]:
```suggestion
We are going to create a `preload` script that exposes the versions of Chrome, Node, and
```

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+
+contextBridge.exposeInMainWorld('versions', {
+  node: () => process.versions.node,
+  chrome: () => process.versions.chrome,
+  electron: () => process.versions.electron
+})
+```
+
+The above code accesses the Node.js `process.versions` object and exposes some of them in
+the renderer process in a `global` object called `versions`.
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:66]:
```suggestion
the renderer process in a [global](https://developer.mozilla.org/en-US/docs/Glossary/Global_object) object called `versions`.
```

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+</html>
+```
+
+We have created a paragraph with `info` as its id, and included `renderer.js`.
+
+:::warning
+The reason why we do not write the JavaScript directly in the HTML between `<script></script>`
+is because we have a Content-Security-Policy that will prevent its execution.
+To know more you can visit [MDN's CSP content][mdn-csp].
+:::
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:144]:
```suggestion
The reason why we do not write the JavaScript directly in the HTML between `<script></script>`
is because we have a Content Security Policy (CSP) that will prevent its execution.
To know more you can visit [MDN's CSP documentation][mdn-csp].
:::
```

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+  combined path string that works across all platforms.
+
+If you are not familiar with Node.js, we recommend you to read [this guide][node-guide].
+:::
+
+At this point, the renderer context has access to `versions`, so let's make that information
+available to the user. Create a new file `renderer.js` that contains the following code:
+
+```js title="renderer.js"
+const information = document.getElementById('information')
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:108]: This doesn't match the ID in the `index.html`
```suggestion
const information = document.getElementById('info')
```

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+Electron into the renderer. To do this, add a new file `preload.js` with the
+following:
+
+```js title="preload.js"
+const { contextBridge } = require('electron')
+
+contextBridge.exposeInMainWorld('versions', {
+  node: () => process.versions.node,
+  chrome: () => process.versions.chrome,
+  electron: () => process.versions.electron
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:61]: Any reason why this these are functions rather than exposed variables?

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+## Main and Renderer process communication
+
+Electron's main process is a Node.js process that has full OS access.
+On top of [Electron's modules][modules], you can also access [Node.js ones][node-api],
+as well as any compatible code downloaded via `npm`.
+
+If you are not familiar with Node.js, we recommend you to first
+read [this guide][node-guide] before continuing.
+
+## Using `contextBridge` and a preload script
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:34]: We should explain the context bridge API a bit more here, and link the context isolation doc.

File: docs/latest/tutorial/tutorial-main-renderer.md [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+
+:::tip Further reading 📚
+This is your last reminder to look at [Electron's Process Model][process-model] to
+better understand all of this!
+:::
+
+This is where attaching a [`preload` script][preload-script] to your
+renderer comes in handy. A `preload` script runs before the renderer process is loaded,
+and has access to both renderer globals (e.g. `window` and `document`) and a Node.js
+environment.
```
Comment (erickzhao) [inline docs/latest/tutorial/tutorial-main-renderer.md:49]: If we're sticking with `sandbox: true`,  we should link the preload section of the Sandboxing doc.

https://www.electronjs.org/docs/latest/tutorial/sandbox#preload-scripts

File: docs/latest/fiddles/tutorial-main-renderer/main.js [hunk 32]
```diff
@@ -0,0 +1,19 @@
+const { app, BrowserWindow } = require('electron')
+const path = require('path')
+
+const createWindow = () => {
+  const win = new BrowserWindow({
+    width: 800,
+    height: 600,
+    webPreferences: {
+      preload: path.join(__dirname, 'preload.js'),
+      sanbox: true
```
Comment (erickzhao) [inline docs/latest/fiddles/tutorial-main-renderer/main.js:10]:
```suggestion
      sandbox: true
```

File: docs/latest/fiddles/tutorial-main-renderer/renderer.js [hunk 33]
```diff
@@ -0,0 +1,2 @@
+const information = document.getElementById('information')
```
Comment (erickzhao) [inline docs/latest/fiddles/tutorial-main-renderer/renderer.js:1]:
```suggestion
const information = document.getElementById('info')
```

**Review by erickzhao — CHANGES_REQUESTED** [review]
Review for pt 3

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 24 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]: You are right.
What about adding a small IPC section right after. Could be in another document or this same one.

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 23 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:25]: I think it can be removed. 

```suggestion
```

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 29 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:61]: The reason was to show it can be custom functions and not just variables. When I was going through the quick start I was wondering if I could do this or not. I could add another property to the object to show a mix and match of options to make it clearer.

File: docs/latest/tutorial/tutorial-scaffolding.md (hunk 16 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-scaffolding.md]: I've updated the `request: "attach"` in case you want to take another look

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 24 above)
Comment (erickzhao) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]: This code review reminded me to pick up writing the IPC doc again. Maybe add a TODO for me?

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 30 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:33]: Done!

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 31 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:48]: Done

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 29 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:60]: I've added a comment explaining we can also expose variables.

File: docs/latest/tutorial/tutorial-main-renderer.md (hunk 24 above)
Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]: I've added a todo at the end of the doc.
Still need to think about a better name 🤔
What about this?

> Exposing features in the renderer process

Comment (erickzhao) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]: Maybe "Using Electron and Node APIs in the renderer" would be a bit more descriptive, not sure if it's too wordy.

Comment (molant) [reply docs/latest/tutorial/tutorial-main-renderer.md:2]:
```suggestion
title: 'Using Electron and Node APIs in the renderer'
```

Sold

File: docs/latest/tutorial/code-signing.md [hunk 34]
```diff
@@ -1,37 +1,49 @@
...
-title: "Code Signing"
-description: "Code signing is a security technology that you use to certify that an app was created by you."
+title: 'Code Signing'
+description: 'Code signing is a security technology that you use to certify that an app was created by you.'
 slug: code-signing
 hide_title: false
 ---
 
-# Code Signing
+:::info Tutorial parts
```
Comment (erickzhao) [inline docs/latest/tutorial/code-signing.md:8]: IMO the full code-signing doc should be separate from the tutorial. I feel like this is too long and has too many options to be part of the step-by-step tutorial.

I'm suggesting two alternative presentations for this info:
* A subsection of "Application Distribution" that references this doc and explains the basics of code-signing.
* Replacing this doc with a step-by-step tutorial on Windows and macOS code-signing in Electron Forge, and leaving the previous code-signing doc alone.

File: docs/latest/tutorial/updates.md [hunk 35]
```diff
@@ -1,11 +1,24 @@
...
-title: "Updating Applications"
+title: 'Updating Your Application'
 description: "There are several ways to update an Electron application. The easiest and officially supported one is taking advantage of the built-in Squirrel framework and Electron's autoUpdater module."
 slug: updates
 hide_title: false
 ---
 
-# Updating Applications
+:::info Tutorial parts
+This is part 7 of the Electron tutorial. The other parts are:
```
Comment (erickzhao) [inline docs/latest/tutorial/updates.md:9]: I think again that the tutorial section should be be streamlined to support a single easy use-case. I think setting up `update.electronjs.org` updates might be a good use-case to extract into the tutorial, but advanced instructions should be in their own doc.

File: docs/latest/tutorial/application-distribution.md [hunk 36]
```diff
@@ -1,56 +1,131 @@
...
+:::tip Further reading 📚
+It is recommended to read their documentation to know how to rebrand and sign the
+executable, change the icons, etc.
+
+You should also read all the documents under the "Distribution" category in the sidebar.
+:::
 
 ## Manual distribution
 
+If you prefer the manual approach, there are 2 ways to distribute your application:
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/latest/tutorial/application-distribution.md:100]: Agree!

**Comment by erickzhao** [comment]
Commandeering this draft since Anton is OOO this week.

File: docs/latest/tutorial/tutorial-2-scaffolding.md [hunk 37]
```diff
@@ -0,0 +1,354 @@
...
+<!--https://www.electronforge.io/guides/developing-with-wsl-->
+
+:::
+
+The interactive `init` command will prompt you to set some fields in your config.
+There are a few rules to follow for the purposes of this tutorial:
+
+- `entry point` should be `main.js`.
+- `author`, `license`, and `description` can be any value, but will be necessary for
+  [app packaging][packaging-distribution] later on.
```
Comment (molant) [inline docs/latest/tutorial/tutorial-2-scaffolding.md:47]: Link is broken, I guess you meant this:

```suggestion
  [app packaging][packaging] later on.
```
